staging/kos/pkg/controllers/subnet/validator.go

```diff
@@ -0,0 +1,524 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package subnet
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+
+	k8scorev1api "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	k8smetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8stypes "k8s.io/apimachinery/pkg/types"
+	k8sutilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	k8swait "k8s.io/apimachinery/pkg/util/wait"
+	k8scache "k8s.io/client-go/tools/cache"
+	k8sworkqueue "k8s.io/client-go/util/workqueue"
+
+	netv1a1 "k8s.io/examples/staging/kos/pkg/apis/network/v1alpha1"
+	kosclientv1a1 "k8s.io/examples/staging/kos/pkg/client/clientset/versioned/typed/network/v1alpha1"
+	netlistv1a1 "k8s.io/examples/staging/kos/pkg/client/listers/network/v1alpha1"
+	kosctlrutils "k8s.io/examples/staging/kos/pkg/controllers/utils"
+)
+
+// TODO: something which makes the validation failure visible to the client
+// (event, Prometheus, status field, all of them, etc...).
+// TODO: some log messages don't print the specific values that yield problems,
+// only what the problem is (e.g. "CIDRs overlap", but no CIDR values). Add such
+// values (this might require changing some functions signatures).
+
+// usable is the value the validator sets a subnet's status.ValidationOutcome to
+// if there are no conflicts with other subnets.
+const usable = "usable"
+
+// conflictsCache holds information for one subnet regarding conflicts with
+// other subnets. It stores three mutually exclusive caches for CIDR-only
```

Comment:
The three lists are not consumed differentially, only their union is consumed.  So I do not see why three distinct lists are being maintained.

Reply:
Also, the lists are being given more complex semantics than they need (as evidenced by, e.g., the checking in line 442).  This cache only need to remember, for subnet X, the subnets Y that were judged to be invalid because of a conflict with subnet X; this is because we only need this cache to know what else to queue when there is a change in X.

Reply:
The three lists ARE consumed differentially. But maybe their names and comments do not make that clear enough. The prerequisite for two subnets S1 and S2 being in conflict is that they have the same VNI. Given that prerequisite, S1 and S2 can be in conflict for two different reasons:

1) Their CIDRs overlap.  
2) They belong to different K8s namespaces.

The first list stores conflicts due to reason 1) only, the second list stores conflicts due to reason 2) only, and the third list stores conflicts due to both 1) and 2). When a subnet Sx is updated, the validator checks what has changed: VNI only, CIDR only or both; it then enqueues the subnets in the appropriate lists only.
This is complex and makes the code hard to reason about wrt using only one list as you're suggesting. The reason I chose to have three lists is that any time an existing subnet is processed after being popped by the queue, there's a live list to the API server, which is bad performance-wise. Say that you have a single list storing all conflicting subnets regardless of the nature of the conflict. Now say that a subnet Sx's CIDR is updated. With the single list solution, you re-process all the existing subnets conflicting with Sx, even those in conflict with it because they have different namespaces, which will still be in conflict. For every such subnet which still exists, there will be a list against the API server which could have been avoided.

Notice that the current solution is not perfect wrt the goal of minimizing useless lists to the API server, because if a subnet Sy in Sx's conflicts cache is updated and ceases to be in conflict with Sx, it's not removed from Sx's cache, which means that if Sx is later updated Sy will be re-processed again with a list against the API server even if there's no need. Handling this case would make for even more complex code.

But maybe you're right about having a single, generic list. One argument in favor of that solution is that conflicts are not the norm, but a rare event caused by a user's mistake. The validator is there only to make sure they do not result in downstream damage. It should not be optimized at the cost of simplicity because in the intended scenario of operations there are few conflicts, which means that even with the single list solution performance will not be compromised too much.

I see three directions from here: keep the code as it is, optimize it further to completely avoid useless lists against the API server, make it simpler by using a single, generic list. What are your thoughts?

Reply:
I read the validator code again and I noticed that:
1) There's an error, subnets in conflict because of overlapping CIDRs should be enqueued again on VNI updates as well, and currently they are not.
2) The comments on the three lists definitely do not make it clear that they are consumed differentially.

Fixes will follow depending on our final choice on whether to keep three separate lists or have a single one.

Reply:
> Also, the lists are being given more complex semantics than they need (as evidenced by, e.g., the checking in line 442). This cache only need to remember, for subnet X, the subnets Y that were judged to be invalid because of a conflict with subnet X; this is because we only need this cache to know what else to queue when there is a change in X.

Assume the cache stored only the conflicts as you suggest, hence a check like the one in line 442 was neither needed nor possible. Now consider the following scenario: subnet Sy is created, the worker processing it does a live list against the API server and detects a conflict with another subnet Sx, hence adds Sy to Sx's conflicts cache. Right after the live list, but before the addition of Sy to Sx's conflicts cache, Sx is updated in such a way to make the conflict disappear, and all subnets in Sx's conflicts cache are enqueued. Sy is added to Sx's cache after the latter enqueuing, hence it will not be reprocessed! In a nutshell, Sy is not marked as usable even if there are no conflicts, and will not be reprocessed until an update to Sx takes place (possibly, never).

Conflicts are not between subnets, but between specific subnets versions. Hence, the conflicts cache should store information about its subnet's version.  Doing this, coupled with a check such as the one in line 442, prevents race conditions like the one above. If a worker processing a subnet Sy detects a conflict with Sx, but the conflicts cache of Sx records a different version of Sx as the one seen by the worker, the conflicts cache is either lagging behind or forging ahead the version of Sx seen by the worker. So the worker requeues Sy, and hopefully when it is processed again Sx's conflicts cache has synced with the latest Sx's version.

Given what I've written above, two questions are still unanswered to have a solution.

The first question is: where do we record the latest version in a subnet's conflicts cache? 

There are two possibilities: 
1) Notification handlers.
2) Worker goroutines.
IMO 2) is better and is what the current code does. 1) means edge-based behavior.
 
Next question is how to model the version of a subnet. I thought about two possibilities:
1) metadata.resourceVersion 
2) CIDR and VNI (the current solution)

2) allows some micro-optimizations which are not possible with 1). For instance, consider the case where a subnet Sx is updated by enlarging its CIDR (i.e., all the old addresses are retained, and new ones are added). There's no need at all to enqueue the subnets in Sx's conflicts cache (and some of those enqueues will lead to live lists against the API server!), but for a worker comparing the metadata.resourceVersion recorded in Sx's conflicts cache with that in the Sx that it got from the informer's cache there's no way to realize that. If instead the comparison is based on CIDR and VNI, the worker can avoid enqueuing subnets uselessly.

Reply:
As you said, conflicts should be rare.  I think we should start with a solution that is simple but correct.  That means an amalgamanted conflict list for each subnet.

We should modify the extension server to initialize's a Subnet's `ObjectMeta` to `time.Now().UnixNano()` and bump the `ObjectMeta.Generation` whenever the Spec changes.

I want a behavioral constraint that we might not have talked enough about.  Once a Subnet is decided to be valid, I want that decision to hold fast until the Subnet is deleted or its `ObjectMeta.Generation` changes.  It is not safe to implement this memory in SubnetStatus; the validation controller needs a local memory in which to remember these decisions.  That local memory needs to be properly initialized from SubnetSpec & Status at startup.  Also, since multiple concurrent validation controllers are possible, and they might reach different decisions, in this case it is safe to mark the Subnet as validated and a controller that did not reach this decision must adopt it if notified of it.  The memory only holds positive decisions.

The conflict memory should map Namespace&Name X to a struct that holds generation G and list L of Namespace&Name, with the meaning that if/when X is ever deleted or advanced to a different generation then the members of L need to be reconsidered.  When, in the course of processing Sy and finding a conflict with Sx of generation Gx, the attempt to add Sy to the conflict list of Sx fails X's conflict cache holds a different generation than Gx; failure causes the processing of Sy to abort and Sy to be requeued.

Reply:
> We should modify the extension server to initialize's a Subnet's ObjectMeta.Generation to time.Now().UnixNano() and bump the ObjectMeta.Generation whenever the Spec changes.

We should do that, but treat `ObjectMeta.Generation` as a counter rather than use `time.Now().UnixNano()`. The accepted semantic of `ObjectMeta.Generation` is that it is monotonically increasing. If we use `time.Now().UnixNano()` that monotonicity might be violated if two updates on the same objects are sent in quick succession to different API servers.

Reply:
> As you said, conflicts should be rare. I think we should start with a solution that is simple but correct. That means an amalgamanted conflict list for each subnet.
Ok, I will change that.

Reply:
I will take for granted the fact that there will be a single conflicts list as opposed to three, and discuss the rest of your solution.

I understand what the design/implementation you're proposing looks like, but less the why. I went through the [API design principles](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/principles.md) again, and I see that having a field such as `status.validationOutcome` violates those principles. First, it pollutes a user-facing API object with an implementation detail irrelevant to the user. Second, it is not reconstructible through observation. Your solution does not solve the first issue as there would still be a `Status.ValidGeneration` field that the user should not see. As for the second issue, I'm not sure, but I'd say no. In the current solution there's no way that  `Status.validationOutcome` can be reconstructed through observation. In the solution you're proposing, a controller could reconstruct that by looking at its conflicts cache, but that would not apply to a newly-created controller. I bet what "reconstructible through observation" means is "a controller must be able to reconstruct the status through observation of the external world", not its internal data structures.

Also, in your comments you not only describe recording if a subnet is invalid in its status as poor (agreed), but as unsafe. Could you provide an example of this? I only see that this choice can lead to some useless processing in certain cases, but not how it leads to incorrect behavior. For instance, in a different comment you wrote: 

> The pattern of keeping implementation state in Status is inferior. In this case it can lead to changes in s.Status.ValidationOutcome that we probably do not want. Consider what happens if a Subnet is processed again before the informer's cache is notified about the update resulting from previous processing.

I assume that the subject is the subnet validator, not some downstream component using subnets. If such re-processing happens again, the subnet version must be the version which was processed when the subnet was marked as valid (usable), otherwise the update with which it was marked as valid would have failed the resource version check. So if such reprocessing happens, it's a duplicate processing which leads to another live list and possibly another update of the subnet, which will fail. This is bad for performance, but correctness is not compromised.

Also, why using `ObjectMeta.Generation` as opposed to `ObjectMeta.resourceVersion`? The latter is already implemented, does everything we need and we would still be treating it opaquely, which is a must.

Finally, you wrote: 

> the conflicts memory only stores entries for positive decisions

Why? If two subnets are in conflict and neither of them is marked as usable, when one of the two is deleted the other should be reconsidered.

Reply:
>  Also, since multiple concurrent validation controllers are possible, and they might reach different decisions, in this case it is safe to mark the Subnet as validated and a controller that did not reach this decision must adopt it if notified of it. The memory only holds positive decisions.

and 

> When, in the course of processing Sy and finding a conflict with Sx of generation Gx, the attempt to add Sy to the conflict list of Sx fails if X's conflict cache holds a different generation than Gx; failure causes the processing of Sy to abort (the conditional reset of Sy's conflict cache, depending on whether Sy's Generation has changed, should happen before evaluating conflicts and persist even if there is an abort here) and Sy to be requeued.

Are both true in the current solution.

Reply:
Validation outcome or ValidGeneration is not a private implementation detail; it is the implementation's response to the desired state.  Part of the implementation's job is to evaluate the desired state and decide which subset is consistent enough for clients to use.

My proposed constancy of decision meets the criterion "Object status must be 100% reconstructable by observation. Any history kept must be just an optimization and not required for correct operation."  My implementation idea of making the controller have a local memory of its positive decisions is not denying that those decisions go into Status, but is meant to cover the scenarios where subsequent processing is done before the controller is notified of that Status update.  To be clear, I think that "by observation" should be taken broadly, and can include observation of the Status of managed objects.

Here is a scenario with a safety violation.  Two conflicting Subnets are created in quick succession.  When the controller processes the first of them, the controller finds no conflict and requests that the first Subnet's Status be marked as validated.  Then, before being notified of that update, the controller processes the first Subnet again.  This time it finds a conflict.  Now that I think about it, this is not a problem if the controller takes no action in this case.  It would be a problem if the controller requests a Status update to mark the Subnet as _not_ valid.

Note that my proposed behavior for Generation is indeed incrementing on changes.  The use of time is only at create time.

The reason for my focus on Generation with this behavior is that it gives me a way to state the desired constancy of decision.  ResourceVersion is modified with any change; I propose that Generation increments only when VNI or CIDR changes, which is exactly what invalidates a previous positive decision.

My think proposal _has_ the property that "If two subnets are in conflict and neither of them is marked as usable, when one of the two is deleted the other should be reconsidered".  The memory of positive decisions has nothing to say about either Subnet in that scenario.  The conflict list for the deleted Subnet should include the remaining Subnet, so the queue worker will requeue the remaining Subnet when processing deletion of the deleted Subnet.

My proposal requires no query of anything in the operation to add an entry to a conflict list, only a comparison of Generation values already in hand.

Reply:
On further reflection, you are right.  Supposing that the validation controller only does writes that are conditional on ResourceVersion match, my idea about a local memory of positive decisions is only about avoiding unnecessary work.  In fact, I think the idea can be simplified and broadened, as follows.

Go ahead and simply keep the memory of positive decisions in Status.

To avoid unnecessary processing work on a given Namespace&Name, keep a memory that holds a ResourceVersion value.  Initialize the value to emtpy.  Whenever the controller gets a successful reply to a write, remember the ResourceVersion that was overwritten (i.e., the old ResourceVersion, not the new one).  Whenever a queue worker starts working on an object, and the informer cache's ResourceVersion for that object matches this memory, the queue worker can safely stop working --- it is guaranteed that there is a coming notification with newer state.  OTOH, if the informer cache holds a different ResourceVersion for the object then this memory can safely be empties; it would also be safe to not change this memory in that situation.

Reply:
Also, there is an alternative to using Generation to characterize and realize the desired constancy behavior.  The desired behavior is something that we define, and we are not restricted to what a controller can do when making that definition.  We can say that once a positive decision has been made for a Subnet with a particular ResourceVersion, VNI, and CIDR, that decision should remain positive until the VNI or CIDR changes or the Subnet is deleted.  One possible choice for representation and implementation is that Status simply holds a `Valid bool` field, and the extension server sets this Status field to false whenever there is a change in desired VNI or CIDR.  This has the virtue of being simpler.  However, it puts part of the implementation in the extension server, rather than leaving the implementation entirely to the validation controller.

Reply:
After a little more thought, I am not so bothered by the latter approach.  Even in the former approach, corresponding considerations go into the definition of the object type and how the extension server handles it.

Reply:
> One possible choice for representation and implementation is that Status simply holds a Valid bool field, and the extension server sets this Status field to false whenever there is a change in desired VNI or CIDR. 

This is what is currently being done, with the exception that Valid is not a boolean but a string, and is therefore set to "" instead of false.

Reply:
Right.

Reply:
Oh but wait, I also proposed using Generation in the conflict list.

Reply:
I think it is safe, and not costly when there are no conflicts, to replace Generation with ResourceVersion in the conflict data structure for a given Subnet.  To recap: for a given Subnet Namespace&Name X we remember a ResourceVersion RVX for X and a list L of Namespace&Name.  When Y is in X's L this means that the last processing of Y found a conflict with X and at that time X's ResourceVersion was RVX.  When processing X and it is being deleted or its ResourceVersion no longer matches RVX, each Y in L is requeued if the informer's local cache for Y does not show `Status.Valid` and then RVX is updated and L is set to empty.  When processing Y and a conflict is found with X having ResourceVersion V then there is an attempt to add Y to X's L.  This attempt fails, causing a requeue of Y, if V != the ResourceVersion currently in X's conflict data.

Reply:
We've agreed on these two changes:
1) The string SubnetStatus.ValidationOutcome will become the bool SubnetStatus.Usable.
2) The three lists in the conflicts cache will become a single, amalgamated list.

On the rest, I'm a bit confused, I'll read again the comments. 
I think the current Validator already does most of what we're looking. The extension server already implements the behavior you described: https://github.com/MikeSpreitzer/kube-examples/blob/make-subnets-validation-correct/staging/kos/pkg/registry/network/subnet/strategy.go#L93-L99.
Likewise, the optimization you're suggesting to avoid duplicate processing after a positive status update is already taken care of in the update notification handler, although it does not rely on resource version, but on the subnet CIDR and VNI: https://github.com/MikeSpreitzer/kube-examples/blob/make-subnets-validation-correct/staging/kos/pkg/controllers/subnet/validator.go#L177-L180.
The current conflicts cache already stores something which is used like generation, and they are owning subnet's VNI and CIDR (notice that their presence is what makes the three separate lists optimization possible --- in fact, when I started the validator it only had one amalgamated list, but the presence of the aforementioned fields made the optimization easy to implement so here it is).

Reply:
> I think it is safe, and not costly when there are no conflicts, to replace Generation with ResourceVersion in the conflict data structure for a given Subnet. To recap: for a given Subnet Namespace&Name X we remember a ResourceVersion RVX for X and a list L of Namespace&Name. When Y is in X's L this means that the last processing of Y found a conflict with X and at that time X's ResourceVersion was RVX. When processing X and it is being deleted or its ResourceVersion no longer matches RVX, each Y in L is requeued if the informer's local cache for Y does not show Status.Valid and then RVX is updated and L is set to empty. When processing Y and a conflict is found with X having ResourceVersion V then there is an attempt to add Y to X's L. This attempt fails, causing a requeue of Y, if V != the ResourceVersion currently in X's conflict data.

This is mostly what the current validator code does. The only difference is that instead of relying on resourceVersion the validator relies on a subnet VNI and CIDR, e.g. the conflicts cache for  a subnet does not store its resource version but its VNI and CIDR.

Reply:
On reconsideration, you're right, the conflict list can be conditioned on VNI&CIDR instead of ResourceVersion.

I deliberately suggested putting the suppression of duplicate work in the queue worker rather than the informer notification handler because the latter is not the only thing that enqueues object references.  But there is no harm in suppressing work in both places.  But I recommend suppressing work based on ResourceVersion because that will not falsely suppress work after VNI or CIDR is changed and then quickly changed back.

Reply:
> I deliberately suggested putting the suppression of duplicate work in the queue worker rather than the informer notification handler because the latter is not the only thing that enqueues object references.

You're right, if a subnet Sx is marked as valid after an update and was previously in the conflicts list of another subnet Sy which is updated/deleted, Sx namespace&name might be enqueued without going through the notification handler, the current code does not handle that.

Reply:
> But I recommend suppressing work based on ResourceVersion because that will not falsely suppress work after VNI or CIDR is changed and then quickly changed back. Or just putting the suppression of duplicate work in everything that enqueues references.

Thinking about it, I too like the resource version solution more. I think it has a more general/understandable semantic, whereas the VNI and CIDR based solution only applies to our specific case. But about this statement: 
> But I recommend suppressing work based on ResourceVersion because that will not falsely suppress work after VNI or CIDR is changed and then quickly changed back

What harm could come from that false suppression of work? It seems to me that it's better performance wise and correctness is not compromised. Although, once again the code is more complex to reason about.

Reply:
So to recap one last time, I'm going to change these things:
1) string SubnetStatus.ValidationOutcome will become bool SubnetStatus.Usable.
2) The three separate conflicts lists will become  a single list.
3) The conflicts cache will record resourceVersion instead of VNI and CIDR. All duplicate work suppression will be based on resourceVersion comparisons. 
4) The IPAM controller and the subnet validator will be moved within a single container/pod, just like the upstream controllers are managed by the controller-manager.

Please let me know if I forgot something.

Reply:
Here is a bad scenario that can happen if work is suppressed based on VNI&CIDR.  Suppose a Subnet X is created with CIDR A and marked valid.  Then suppose its CIDR is changed to something completely different, B.  Then suppose its CIDR is changed back to A.  Then suppose the controller processes X a second time.  There is no guarantee that the controller ever sees the state with CIDR == B.  If seeing CIDR==A a second time causes no processing, then the controller will fail to validate the Subnet.

Reply:
Regarding your recap: I think you put together into item 3 two ideas that I have been imagining are separate.  One is the conflict cache.  It holds VNI&CIDR and list of Namespace&Name.  The other is the duplicate work suppression; that is based on a memory of ResourceVersion.

Reply:
You're right both about the bad scenario and the fact that I had merged the two ideas together.
I'm at work on the changes.